pareto/strategic_water_management/strategic_produced_water_optimization.py

```diff
@@ -4069,6 +4159,575 @@ def DisposalDestinationDeliveriesRule(model, k, t):
     return model
 
 
+def water_quality(model, df_parameters, df_sets):
+    # region Fix solved Strategic Model variables
+    for var in model.component_objects(Var):
+        for index in var:
+            # Check if the variable is indexed
+            if index is None:
+                # Check if the value can reasonably be assumed to be non-zero
+                if abs(var.value) > 0.0000001:
+                    var.fix()
+                # Otherwise, fix to 0
+                else:
+                    var.fix(0)
+            elif index is not None:
+                # Check if the value can reasonably be assumed to be non-zero
+                if var[index].value and abs(var[index].value) > 0.0000001:
+                    var[index].fix()
+                # Otherwise, fix to 0
+                else:
+                    var[index].fix(0)
+    # endregion
+
+    # Create block for calculating quality at each location in the model
+    model.quality = Block()
+
+    # region Add sets, parameters and constraints
+
+    # Crate a set for Completions Pad storage by appending "-storage" to each item in the CompletionsPads Set
+    df_sets["CompletionsPadsStorage"] = [
+        p + "-storage" for p in df_sets["CompletionsPads"]
+    ]
+    model.quality.s_CP_Storage = Set(
+        initialize=df_sets["CompletionsPadsStorage"],
+        doc="Completions Pad Storage Tanks",
+    )
+
+    # Create a set for water quality at Completions Pads intermediate flows (i.e. the blended trucked and piped water to pad)
+    df_sets["CompletionsPadsIntermediate"] = [
+        p + "-intermediate" for p in df_sets["CompletionsPads"]
+    ]
+    model.quality.s_CP_Intermediate = Set(
+        initialize=df_sets["CompletionsPadsIntermediate"],
+        doc="Completions Pad Intermediate Flows",
+    )
+
+    # Create a set of locations to track water quality over
+    model.quality.s_WQL = Set(
+        initialize=(
+            model.s_L | model.quality.s_CP_Storage | model.quality.s_CP_Intermediate
+        ),
+        doc="Locations with tracked water quality ",
+    )
+
+    # Quality at pad
+    model.quality.p_nu_pad = Param(
+        model.s_P,
+        model.s_W,
+        default=0,
+        initialize=df_parameters["PadWaterQuality"],
+        doc="Water Quality at pad [mg/L]",
+    )
+    # Quality of Sourced Water
+    model.quality.p_nu_freshwater = Param(
+        model.s_F,
+        model.s_W,
+        default=0,
+        initialize=1000,
```

Comment:
Another good point. We decided to set this value to 0, a better estimate for water components like lithium. 1000 mg/L vs  0 mg/L is not too different for TDS because our ranges are ~ 150,000.